runtime/lua/vim/iter.lua

```diff
@@ -179,6 +194,59 @@ function ListIter.filter(self, f)
   return self
 end
 
+--- Flattens an |list-iterator| pipeline, un-nesting nested values.
+--- The flattened list will exclude nil, and any values with non-numeric keys
+---
+--- Example:
+---
+--- ```lua
+--- vim.iter({1, {2}, {{3}}}):flat():totable()
+--- ```
+--- (returns `{1, 2, {3}}`)
+---
+--- ```lua
+--- vim.iter({1, {a = 2}, {nil, {nil}}, {{3}}}):flat():totable()
+--- ```
+--- (returns `{1, {}, 3}`)
+---
+--- ```lua
+--- vim.iter({[2] = 2}):flat():totable()
+--- vim.iter({nil, 2})
+---
+--- ```
+--- Both give an error because the iterator being flattened is not list-like
+--- (the indices in the source table are not consecutive)
+---
+---@param depth number|nil Depth to which iterator pipeline should be flattened
+---                        (defaults to 1)
+---@return Iter
+function Iter.flat(self, depth) -- luacheck: no unused args
+  error('flat() requires a list-like table')
+  return self
+end
+
+---@private
+function ListIter.flat(self, depth)
```

Comment:
```suggestion
function ListIter.flatten(self, depth)
```

Reply:
@gpanders do you mind sticking with `flat` ?